Api routes

IlanSirisky opened the pull request.

IlanSirisky committed:
DB setup with initial app routes

IlanSirisky committed:
Update to DB Tables

IlanSirisky committed:
Added search-pokemon route

Added search-pokemon route with search and sort by functionality

IlanSirisky committed:
Updated search pokemons and getPokemonById endpoints

IlanSirisky committed:
small fix

IlanSirisky committed:
Changed code to run with prisma

IlanSirisky committed:
add validation middleware

IlanSirisky committed:
Refactored data structure

IlanSirisky committed:
Fixed random pokemon endpoint

IlanSirisky committed:
Delete types.json

IlanSirisky committed:
Added error handling middleware

Improved code readability

IlanSirisky committed:
Update pagination handling in backend

IlanSirisky committed:
Updated searchPokemon validation

IlanSirisky committed:
Separated controller to handler file

IlanSirisky committed:
Moved fight logic to backend

IlanSirisky committed:
Moved catch endpoint to battle controller

oranroz11 commented on package.json:
```diff
@@ -11,16 +11,20 @@
     "@types/node": "^20.14.0",
     "@types/pg": "^8.11.6",
     "nodemon": "^3.1.2",
+    "prisma": "^5.15.0",
```
remove ^

oranroz11 commented on prisma/migrations/20240606072714_init/migration.sql:
```diff
@@ -0,0 +1,42 @@
+    "description" TEXT NOT NULL,
+    "image" TEXT NOT NULL,
+
+    CONSTRAINT "Pokemon_pkey" PRIMARY KEY ("id")
```
please change id - `"id" SERIAL PRIMARY KEY,`

oranroz11 commented on prisma/migrations/20240606072714_init/migration.sql:
```diff
@@ -0,0 +1,42 @@
+    "weight" TEXT NOT NULL,
+    "ability" TEXT[],
+    "types" TEXT[]
+);
```
`CREATE TABLE "Profile" (
    "pokemon_id" INTEGER NOT NULL,
    "height" TEXT NOT NULL,
    "weight" TEXT NOT NULL,
    "ability" TEXT[],
    "types" TEXT[],
    CONSTRAINT "Profile_pokemon_id_fkey" FOREIGN KEY ("pokemon_id")
        REFERENCES "Pokemon"("id")
        ON DELETE RESTRICT
        ON UPDATE CASCADE
);`

oranroz11 commented on prisma/migrations/20240606072714_init/migration.sql:
```diff
@@ -0,0 +1,42 @@
+    "power_level" INTEGER NOT NULL,
+
+    CONSTRAINT "BaseStats_pkey" PRIMARY KEY ("pokemon_id")
+);
```
CREATE TABLE "BaseStats" (
    "pokemon_id" INTEGER PRIMARY KEY,
    "hp" INTEGER NOT NULL,
    "attack" INTEGER NOT NULL,
    "defense" INTEGER NOT NULL,
    "sp_attack" INTEGER NOT NULL,
    "sp_defense" INTEGER NOT NULL,
    "speed" INTEGER NOT NULL,
    "power_level" INTEGER NOT NULL,
    CONSTRAINT "BaseStats_pokemon_id_fkey" FOREIGN KEY ("pokemon_id")
        REFERENCES "Pokemon"("id")
        ON DELETE RESTRICT
        ON UPDATE CASCADE
);

oranroz11 commented on prisma/migrations/20240606072714_init/migration.sql:
```diff
@@ -0,0 +1,42 @@
+ALTER TABLE "Profile" ADD CONSTRAINT "Profile_pokemon_id_fkey" FOREIGN KEY ("pokemon_id") REFERENCES "Pokemon"("id") ON DELETE RESTRICT ON UPDATE CASCADE;
+
+-- AddForeignKey
+ALTER TABLE "BaseStats" ADD CONSTRAINT "BaseStats_pokemon_id_fkey" FOREIGN KEY ("pokemon_id") REFERENCES "Pokemon"("id") ON DELETE RESTRICT ON UPDATE CASCADE;
```
delete

oranroz11 commented on prisma/schema.prisma:
```diff
@@ -0,0 +1,41 @@
+  weight     String
+  ability    String[]
+  types      String[]
+  Pokemon    Pokemon  @relation(fields: [pokemon_id], references: [id])
```
you miss the index

IlanSirisky replied:
did you mean this?
  @@index([pokemon_id])

oranroz11 commented on src/controllers/battleController.ts:
```diff
@@ -0,0 +1,48 @@
+      Number(opponentPokemonId)
+    );
+    return res.status(200).json(fightState);
+  } catch (error: any) {
```
create type for error

oranroz11 commented on src/controllers/battleController.ts:
```diff
@@ -0,0 +1,48 @@
+    );
+    return res.status(200).json(fightState);
+  } catch (error: any) {
+    return res.status(500).json({ message: error.message });
```
i would create a responseBuilder that get the message and error code and create the response - for error and success responses

oranroz11 commented on src/controllers/battleController.ts:
```diff
@@ -0,0 +1,48 @@
+  const { playerPokemonId, opponentPokemonId } = req.body;
+
+  try {
+    const fightState = await startFightHandler(
```
please export the handler file and then call the function like that :

battleService.startFight(...)

oranroz11 commented on src/controllers/battleController.ts:
```diff
@@ -0,0 +1,48 @@
+  } catch (error: any) {
+    return res.status(400).json({ message: error.message });
+  }
+};
```
same for all function as the code review in startFight

oranroz11 commented on src/controllers/pokemonController.ts:
same review as i wrote in battleController

oranroz11 commented on src/controllers/pokemonController.ts:
```diff
@@ -0,0 +1,133 @@
+    const pokemons = await getAllPokemonsHandler();
+    res.json(pokemons);
+  } catch (err) {
+    next(err);
```
if the next is for a middlware so please fix all other functions in the controllers, please keep the code consistent

oranroz11 commented on src/controllers/pokemonController.ts:
```diff
@@ -0,0 +1,133 @@
+
+    // Get a random Pokemon
+    const pokemon = await getRandomPokemonHandler(isOwnedBoolean, randomOffset);
+    if (!pokemon) {
```
all the logic should be in the handler

oranroz11 commented on src/controllers/pokemonController.ts:
```diff
@@ -0,0 +1,133 @@
+) => {
+  const {
+    isOwned,
+    q,
```
better naming, what q means?

oranroz11 commented on src/controllers/pokemonController.ts:
```diff
@@ -0,0 +1,133 @@
+    // Calculate skip and take for pagination
+    const skip =
+      (parseInt(page as string, 10) - 1) * parseInt(limit as string, 10);
+    const take = parseInt(limit as string, 10);
```
move to handler

oranroz11 commented on src/handlers/battleHandler.ts:
```diff
@@ -0,0 +1,99 @@
+  opponentPokemon: null as IPokemonData | null,
+  playerCurrentHp: 0,
+  opponentCurrentHp: 0,
+};
```
create type

oranroz11 commented on src/handlers/battleHandler.ts:
```diff
@@ -0,0 +1,99 @@
+};
+
+// Start a fight
+export const startFightHandler = async (
```
delete handler from the name after you make the changes in the controller

oranroz11 commented on src/handlers/battleHandler.ts:
```diff
@@ -0,0 +1,99 @@
+  opponentPokemonId: number
+) => {
+  const playerPokemon = await getPokemonByIdHandler(playerPokemonId);
+  const opponentPokemon = await getPokemonByIdHandler(opponentPokemonId);
```
handlers shouldnt call each other

oranroz11 commented on src/handlers/battleHandler.ts:
```diff
@@ -0,0 +1,99 @@
+  const catchSuccess = Math.random() < catchRate;
+
+  if (catchSuccess) {
+    const updatedPokemon = await catchPokemonHandler(
```
handlers shouldnt call each other

oranroz11 commented on src/handlers/battleHandler.ts:
```diff
@@ -0,0 +1,99 @@
+      message: "Caught the Pokémon!",
+      caught: true,
+      pokemon: updatedPokemon,
+    };
```
add type

oranroz11 commented on src/handlers/battleHandler.ts:
```diff
@@ -0,0 +1,99 @@
+    return {
+      message: "Failed to catch the Pokémon.",
+      caught: false,
+    };
```
add type

oranroz11 commented on src/handlers/pokemonHandler.ts:
all prisma functions should be model file!!!!

oranroz11 commented on src/handlers/pokemonHandler.ts:
```diff
@@ -0,0 +1,79 @@
+    where: { isOwned: true },
+    include: includeRelations,
+  });
+};
```
merge to one function that isOwned is optional

oranroz11 commented on src/handlers/pokemonHandler.ts:
```diff
@@ -0,0 +1,79 @@
+    data: { isOwned: true },
+    include: includeRelations,
+  });
+};
```
in the model file please call the function updatePokemon in case that the only attribute that the user can update is the owner - call the function updateOwnerPokemon

oranroz11 commented on src/handlers/pokemonHandler.ts:
```diff
@@ -0,0 +1,79 @@
+    include: includeRelations,
+  });
+
+  const totalCount = await prisma.pokemon.count({ where });
```
merge the function with the other two function from the top file.

oranroz11 commented on src/middleware/errorMiddleware.ts:
```diff
@@ -0,0 +1,13 @@
+import { Request, Response, NextFunction } from "express";
+
+const errorHandler = (
+  err: any,
```
add type, change err to error

oranroz11 commented on src/middleware/errorMiddleware.ts:
```diff
@@ -0,0 +1,13 @@
+  res: Response,
+  next: NextFunction
+) => {
+  console.error(err);
```
delete

oranroz11 commented on src/routes/battleRoutes.ts:
```diff
@@ -0,0 +1,12 @@
+const router = Router();
+
+/*** POST ***/
+router.post('/start-fight', startFight);
+router.post('/player-attack', playerAttack);
+router.post('/opponent-attack', opponentAttack);
+router.post('/catch', catchPokemon);
```
why they are post? is there body?

oranroz11 commented on src/routes/pokemonRoutes.ts:
```diff
@@ -0,0 +1,23 @@
+const router = Router();
+
+/*** GET ***/
+router.get("/pokemons", getAllPokemons);
+router.get("/pokemons/owned", getOwnedPokemons);
+router.get("/pokemons/:id", validatePokemonId, getPokemonById);
```
you can add the pokemons in the index.ts - app.use("/api/pokemons", pokemonRoutes);

oranroz11 commented on src/types/pokemonType.ts:
```diff
@@ -0,0 +1,48 @@
+  isOwned: boolean;
+  description: string;
+  image: string;
+  profile: Profile | null;
+  baseStats: BaseStats | null;
```
profile?: Profile;
  baseStats?: BaseStats;

IlanSirisky replied:
Added the "?" but cant remove the " | null".
Yuval said to leave it as it is.

If i remove i get this error:
Type '({ profile: { pokemon_id: number; height: string; weight: string; ability: string[]; types: string[]; } | null; baseStats: { pokemon_id: number; hp: number; attack: number; defense: number; sp_attack: number; sp_defense: number; speed: number; power_level: number; } | null; } & { ...; })[]' is not assignable to type 'IPokemonData[]'.
  Type '{ profile: { pokemon_id: number; height: string; weight: string; ability: string[]; types: string[]; } | null; baseStats: { pokemon_id: number; hp: number; attack: number; defense: number; sp_attack: number; sp_defense: number; speed: number; power_level: number; } | null; } & { ...; }' is not assignable to type 'IPokemonData'.
    Types of property 'profile' are incompatible.
      Type '{ pokemon_id: number; height: string; weight: string; ability: string[]; types: string[]; } | null' is not assignable to type 'Profile | undefined'.
        Type 'null' is not assignable to type 'Profile | undefined'.

IlanSirisky committed:
cr fixes

Added pokemonModel, responseBuilder.
Refactored functions hierarchy to controller - handler - model.
moved logic to handler files.
updated prisma schema and migration file.

IlanSirisky committed:
Add start fight validation

IlanSirisky committed:
Merged getPokemons into searchPokemons

IlanSirisky committed:
Update pokemonController.ts
